core/arch/arm/kernel/user_ta.c

```diff
@@ -687,6 +689,39 @@ static TEE_Result add_deps(struct user_ta_ctx *utc __unused,
 
 #endif
 
+#ifdef CFG_TA_ASLR
+static size_t aslr_offset(size_t min, size_t max)
+{
+	uint32_t rnd32 = 0;
+	size_t rnd = 0;
+
+	assert(min <= max);
+	if (max > min) {
+		if (crypto_rng_read(&rnd32, sizeof(rnd32)))
+			return 0;
+		rnd = rnd32 % (max - min);
+	}
+	return (min + rnd) * CORE_MMU_USER_CODE_SIZE;
+}
+
+static vaddr_t get_stack_va_hint(void)
+{
+	vaddr_t base = 0;
+
+	core_mmu_get_user_va_range(&base, NULL);
+
+	/* 1 spare page for the ---R-X page + 1 guard page */
+	return base + aslr_offset(2 + CFG_TA_ASLR_MIN_OFFSET_PAGES,
```

Comment:
The `2` actually depends on what's happening in `map_kinit()`.

It would be better to check for next available VA in `utc->vm_info` if it's available or skip adding `2` if absent.

Reply:
> It would be better to check for next available VA in `utc->vm_info` if it's available or skip adding `2` if absent.

Hmmm... not sure I get it :-/

If I read the first available VA from `utc->vm_info` like so:
```C
r = TAILQ_LAST(&utc->vm_info->regions, vm_region_head);
if (r) {
    va = r->va + r->size; // <==== HERE
} else {
    core_mmu_get_user_va_range(&va, NULL);
}
```
...this VA is rejected by `vm_map()`, apparently due to `pad = SMALL_PAGE_SIZE` [here](https://github.com/OP-TEE/optee_os/blob/3.3.0/core/arch/arm/mm/tee_mmu.c#L61).

What is the reason for this padding? Should I just add one page to `va` in the first leg of the `if`?

Reply:
The padding is to separate privileged mapping from user mode mapping, it seemed like adding a page here was appropriate.

Adding `SMALL_PAGE_SIZE` should take care of that, I don't like depending on internals of `select_va_in_range()` here, but I'm afraid the alternative is even worse. So where you're adding `SMALL_PAGE_SIZE` an explanation is needed. Something like:
```
/*
 * Adding an empty page to separate TA mappings from already present mappings with
 * TEE_MATTR_PERMANENT to satisfy select_va_in_range()
 */

Reply:
OK thanks.